src/database/handler.py

```diff
@@ -0,0 +1,114 @@
+# Standard Library
+from logging import getLogger
+from typing import List
+
+# Third Party
+from sqlalchemy import create_engine
+from sqlalchemy.orm import sessionmaker
+from sqlalchemy_utils import create_database, database_exists
+
+# Local
+from src.common import settings
+from src.common.enums import DocumentStatus
+from src.common.utils import Singleton
+from src.database.models import Base, Document, Page
+
+logger = getLogger(__name__)
+
+
+class DatabaseHandler(metaclass=Singleton):
+    # (TODO: make this class abstract and implement specific handlers)
+
+    def __init__(self):
+        # create PostgreSQL engine to connect to the database
+        self.engine = create_engine(
+            f'postgresql://{settings.DATABASE_USERNAME}:{settings.DATABASE_PASSWORD}@{settings.DATABASE_HOST}:{settings.DATABASE_PORT}/{settings.DATABASE_NAME}',
+            pool_pre_ping=True)
+
+        # create the database if does not exist
+        if not database_exists(self.engine.url):
+            create_database(self.engine.url)
+
+        # connect to the database
+        self.engine.connect()
+
+        # create tables if do not exist
+        Base.metadata.create_all(self.engine, checkfirst=True)
+
+        # create a configured "Session" class
+        self.Session = sessionmaker(bind=self.engine)
+
+    def close(self):
+        # TODO: close database connection
+        pass
+
+    def get_document(self, document_id: str) -> Document:
```

Comment:
So the reason of having asynchronous routes (or any other code) is to be able to give up control in case we wait for some kind of I/O operation to return the result... to allow other part of the code to continue meanwhile if possible. Having sychronous database communication means that we lose this benefit. SQLAlchemy allows to initialize asynchronous engine and respective methods, would make sense to use it (I forgot about it).

Reply:
Well, we would lose that benefit and we would block all other coroutines from getting the control until the SQL query is finished. With some nasty queries (taking couple of seconds) that worker could not work at all. Actually, Andrew Godwin (the one who brought the async to Django) has a really nice slide about that https://speakerdeck.com/andrewgodwin/just-add-await-retrofitting-async-into-django?slide=22.

I'm not aware of SQLAlchemy being available for the usage with async.

From what I found, FastAPI documentation recommends to use only the SQLAlchemy core for generating the SQL queries and `encode/databases` for the async database connection https://fastapi.tiangolo.com/advanced/async-sql-databases/. Or there are other ORMs like GINO (https://github.com/python-gino/gino) or peewee in combination with peewee-async (https://peewee-async.readthedocs.io/en/latest/).

Btw I got to know this only when Django 3.0 was released and I started to get `SynchronousOnlyOperation` exceptions (https://docs.djangoproject.com/en/3.1/ref/exceptions/#django.core.exceptions.SynchronousOnlyOperation) and that was a lot later compared to the creation of the async code 👼 .

Reply:
I remember guys responsible for persistence service in our solution to talk about async SQLAlchemy... so when I looked here: https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html I unfortunately did not check the release. Maybe they actually used some other package like this: https://pypi.org/project/async-sqlalchemy/, not sure. However I agree that it would need to be solve somehow.

Otherwise, thanks for the extra materials, will check it once I have more time. :)